### PR TITLE
Global floating IP block reservation

### PR DESCRIPTION
**Pull request by t0mk**
this PR extends the IP reservation resource to allow global IPv4

### CODE REVIEW

**Comment by t0mk** [comment]
This is ready. I checked that IP address blocks created with previous provider version will not generate diff.

File: packet/datasource_packet_precreated_ip_block.go [hunk 1]
```diff
@@ -78,13 +57,32 @@ func dataSourcePacketReservedIPBlockRead(d *schema.ResourceData, meta interface{
 	}
 	ipv := d.Get("address_family").(int)
 	public := d.Get("public").(bool)
-	facility := d.Get("facility").(string)
+	global := d.Get("global").(bool)
```
Comment (johnstudarus) [inline packet/datasource_packet_precreated_ip_block.go]: Can something be a global but not public? Is that a worthwhile check?

File: packet/datasource_packet_precreated_ip_block.go [hunk 2]
```diff
@@ -78,13 +57,32 @@ func dataSourcePacketReservedIPBlockRead(d *schema.ResourceData, meta interface{
...
+
+	fval, fok := d.GetOk("facility")
+	if fok && global {
+		return fmt.Errorf("You can't specify facility for global IP block - addresses from global blocks can be assigned to devices in any facility")
```
Comment (johnstudarus) [inline packet/datasource_packet_precreated_ip_block.go]: "...assigned to devices across several facilities. For a single facility please use a local IP block."

Better explains that global IPs are to be used by many devices across may facilities.

Comment (johnstudarus) [reply packet/datasource_packet_precreated_ip_block.go]: I'd move the "if fok && global..." check into the "if fok" block below for better reading flow.

File: packet/resource_packet_reserved_ip_block.go [hunk 3]
```diff
@@ -101,33 +131,58 @@ func resourcePacketReservedIPBlockCreate(d *schema.ResourceData, meta interface{
...
+		return err
+	}
+	d.Set("type", typ)
 	d.Set("public", reservedBlock.Public)
 	d.Set("management", reservedBlock.Management)
 	d.Set("manageable", reservedBlock.Manageable)
 	if reservedBlock.AddressFamily == 4 {
 		d.Set("quantity", ipv4CIDRToQuantity[reservedBlock.CIDR])
 	} else {
 		// In Packet, reserved IPv6 block is allocated when device is run in a proejct.
```
Comment (johnstudarus) [inline packet/resource_packet_reserved_ip_block.go]: ^proejct^project

File: website/docs/r/reserved_ip_block.html.markdown [hunk 4]
```diff
@@ -30,22 +31,32 @@ resource "packet_reserved_ip_block" "two_elastic_addresses" {
...
+* `facility` - (Optional) Facility where to allocate the public IP address block, makes sense only for type==public_ipv4
+* `type` - (Optional) Either "global_ipv4" or "public_ipv4", defaults to "public_ipv4" for backward compatibility
+
 
 ## Attributes Reference
 
 The following attributes are exported:
 
-* `facility` - The facility where the addresses are
+* `facility` - The facility where the block was allocated, empty for global blocks
```
Comment (johnstudarus) [inline website/docs/r/reserved_ip_block.html.markdown]: ^empty^must be empty

File: packet/datasource_packet_precreated_ip_block.go (hunk 2 above)
Comment (t0mk) [reply packet/datasource_packet_precreated_ip_block.go]: Thanks, I like your error message better.

If I'd move it to the `if fok {` block, I'd need to check the `global` flag (and thus assert the same conditoin) twice - secondly in the `} else {' block for `!global`.
This way it's checked once, deduping and making the assert quite explicit, I think.

File: packet/datasource_packet_precreated_ip_block.go (hunk 1 above)
Comment (t0mk) [reply packet/datasource_packet_precreated_ip_block.go]: At the moment there can't be any IP blocks which are both global and not-public. A "global" VPN would be a case for non-public global block. I will add the check.

File: website/docs/r/reserved_ip_block.html.markdown (hunk 4 above)
Comment (t0mk) [reply website/docs/r/reserved_ip_block.html.markdown]: I agree that the facility-global_ipv4 constraint should be written explicitly. I added it to the "Argument Reference".

In the "Attributes Reference" section, I think it's proper to only _declare_ that the facility _is_ empty for global_ipv4.

**Comment by t0mk** [comment]
@patrickdevivo @johnstudarus thanks for review. I run the acceptance tests and some are flaking, I think it's because something is wrong inside the Packet API. I will try tomorrow and if it goes through, I will merge.

**Comment by t0mk** [comment]
I removed the commit disabling SpotMarketRequest test, now running the suite again.

**Comment by t0mk** [comment]
the tests pass now, merging

**Pull request merged**